src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/BatchFixAllProvider.cs

```diff
@@ -228,47 +229,97 @@ public virtual string GetFixAllTitle(FixAllContext fixAllContext)
                     case FixAllScope.Document:
                         if (document != null && !generatedCodeServices.IsGeneratedCode(document))
                         {
-                            var diagnostics = await fixAllContext.GetDocumentDiagnosticsAsync(document).ConfigureAwait(false);
-                            var kvp = SpecializedCollections.SingletonEnumerable(KeyValuePair.Create(document, diagnostics));
+                            var documentDiagnostics = await fixAllContext.GetDocumentDiagnosticsAsync(document).ConfigureAwait(false);
+                            var kvp = SpecializedCollections.SingletonEnumerable(KeyValuePair.Create(document, documentDiagnostics));
                             return ImmutableDictionary.CreateRange(kvp);
                         }
 
                         break;
 
                     case FixAllScope.Project:
-                        documentsToFix = project.Documents;
+                        projectsToFix = ImmutableArray.Create(project);
+                        allDiagnostics = await fixAllContext.GetAllDiagnosticsAsync(project).ConfigureAwait(false);
                         break;
 
                     case FixAllScope.Solution:
-                        documentsToFix = project.Solution.Projects
+                        projectsToFix = project.Solution.Projects
                             .Where(p => p.Language == project.Language)
-                            .SelectMany(p => p.Documents);
+                            .ToImmutableArray();
+
+                        var diagnostics = new ConcurrentBag<Diagnostic>();
+                        var options = new ParallelOptions() { CancellationToken = fixAllContext.CancellationToken };
+                        Parallel.ForEach(projectsToFix, options, projectToFix =>
+                        {
+                            fixAllContext.CancellationToken.ThrowIfCancellationRequested();
+
+                            var projectDiagnostics = fixAllContext.GetAllDiagnosticsAsync(projectToFix).WaitAndGetResult(fixAllContext.CancellationToken);
```

Comment:
any reason not using the Task[] and WhenAll pattern we are using other places? this method itself is async so no reason to use Parallel (synchronous one) now.